core/src/main/scala/org/apache/spark/deploy/history/ApplicationCache.scala

```diff
@@ -0,0 +1,579 @@
+  }
+
+  override def toString: String = {
+    appId + attemptId.map { id => s"/$id" }.getOrElse("")
+  }
+}
+
+private[history] class CacheMetrics(prefix: String) extends Source {
+
+  /* metrics: counters and timers */
+  val lookupCount = new Counter()
+  val lookupFailureCount = new Counter()
+  val evictionCount = new Counter()
+  val loadCount = new Counter()
+  val loadTimer = new Timer()
+  val updateProbeCount = new Counter()
+  val updateProbeTimer = new Timer()
+  val updateTriggeredCount = new Counter()
+
+  /** all the counters: for registration and string conversion. */
+  private val counters = Seq(
+    ("lookup.count", lookupCount),
+    ("lookup.failure.count", lookupFailureCount),
+    ("eviction.count", evictionCount),
+    ("load.count", loadCount),
+    ("update.probe.count", updateProbeCount),
+    ("update.triggered.count", updateTriggeredCount))
+
+  /** all metrics, including timers */
+  private val allMetrics = counters ++ Seq(
+    ("load.timer", loadTimer),
+    ("update.probe.timer", updateProbeTimer))
+
+  /**
+   * Name of metric source
+   */
+  override val sourceName = "ApplicationCache"
+
+  override def metricRegistry: MetricRegistry = new MetricRegistry
+
+  /**
+   * Startup actions.
+   * This includes registering metrics with [[metricRegistry]]
+   */
+  private def init(): Unit = {
+    allMetrics.foreach(e =>
+      metricRegistry.register(MetricRegistry.name(prefix, e._1), e._2))
+  }
+
+  override def toString: String = {
+    val sb = new StringBuilder()
+    counters.foreach { e =>
+      sb.append(e._1).append(" = ").append(e._2.getCount).append('\n')
+    }
+    sb.toString()
+  }
+}
+
+/**
+ * API for cache events. That is: loading an APP UI; probing for it changing, and for
+ * attaching/detaching the UI to and from the Web UI.
+ */
+private[history] trait ApplicationCacheOperations {
+
+  /**
+   * Get the application UI.
+   * @param appId application ID
+   * @param attemptId attempt ID
+   * @return If found, the Spark UI and any history information to be used in the cache
+   */
+  def getAppUI(appId: String, attemptId: Option[String]): Option[(SparkUI, Long, Option[Any])]
+
+  /**
+   * Attach a reconstructed UI.
+   * @param appId application ID
+   * @param attemptId attempt ID
+   * @param ui UI
+   * @param completed flag to indicate that the UI has completed
+   */
+  def attachSparkUI(appId: String,
+      attemptId: Option[String],
+      ui: SparkUI,
+      completed: Boolean): Unit
+
+  /**
+   * Detach a Spark UI.
+   *
+   * @param ui Spark UI
+   */
+  def detachSparkUI(appId: String, attemptId: Option[String], ui: SparkUI): Unit
+
+  /**
+   * Probe for an update to an (incompleted) application.
+   * @param appId application ID
+   * @param attemptId optional attempt ID
+   * @param updateTimeMillis time in milliseconds to use as the threshold for an update.
+   * @param data any other data the operations implementation can use to determine age
```

Comment:
Lemme put my complaint another way: FsHistoryProviderL826 has `data.getOrElse(-1L).asInstanceOf[Long]`.  That says to me that `data` really has to be an `Option[Long]`.  If its not, it will blow up at runtime.  So even though the api says you can return an `Option[Any]`, you can't actually code to that.  Its the same issue if there is one common parent class -- the api woudl suggest that you can return any subclass, but in fact you need to return a specific one.  (Though it would be make it a bit more clear what the field meant, at least.)

Taking another look, I guess that `data` is coming from `FsHistoryProvider`, and making its way back to `FsHistoryProvider` unmodified.  So that is why some other provider could have another type in there, which they use similarly?  If that is the use, then it would be properly expressed with a generic.  But also, I'd say that maybe we should just hold off on that for now?  I don't see a point in complicating the api now for hypothetical uses down the line.